### PR TITLE
router: Support regular expression-valued paths

### PR DESCRIPTION
**Pull request by elyscape**
This provides support in the HAProxy router for routes with paths that are treated as regular expressions. Routes are flagged as having path regular expressions by having an annotation of name `haproxy.router.openshift.io/path-regexp` with a truthy value. This also adds support to the `oc` CLI for creating routes with regexp-valued paths.

Resolves: #17156

### CODE REVIEW

**Comment by elyscape** [comment]
Pinging @pweil- to make sure this didn't fall through the cracks.

**Comment by smarterclayton** [comment]
@openshift/sig-networking

File: pkg/router/template/template_helper.go [hunk 1]
```diff
@@ -97,7 +97,7 @@ func genSubdomainWildcardRegexp(hostname, path string, exactPath bool) string {
 }
 
 // Generate a regular expression to match route hosts (and paths if any).
-func generateRouteRegexp(hostname, path string, wildcard bool) string {
+func generateRouteRegexp(hostname, path string, pathRegexp bool, wildcard bool) string {
```
Comment (rajatchopra) [inline pkg/router/template/template_helper.go]: At the very least, the existing function cannot be modified for backward compatibility reasons.

Comment (elyscape) [reply pkg/router/template/template_helper.go]: I’ll add a second function, then.

**Comment by elyscape** [comment]
The regex itself comes from the value of path. The annotation simply indicates whether the path should be interpreted as a regex or as a literal.

**Comment by rajatchopra** [comment]
Right, and treating it as non literal may have security implications.

**Comment by elyscape** [comment]
Gotcha. Yeah, that makes sense; someone could craft a pathological regex designed to DoS the router. So then the question is how to deal with that.

File: pkg/router/template/template_helper.go (hunk 1 above)
Comment (elyscape) [reply pkg/router/template/template_helper.go]: Handled. The new function is called `generateRouteRegexp2`. `generateRouteRegexp` calls out to the new function with `pathRegexp` set to `false`, preserving the current behavior.

**Comment by elyscape** [comment]
Rebased onto master and updated to provide backward compatibility. Is there anything I can do to help with the security implications?

**Comment by elyscape** [comment]
@rajatchopra Should I rebase this? Is there anything I can do to help with the security review?

**Comment by rajatchopra** [comment]
Even though we cannot find a string that will fail the security here, the concern is that we will be at the mercy of haproxy parser for possibility of such a string in this scenario.
With that in mind, unfortunately we cannot allow this implementation.

**Pull request closed without merging**